infix/lexer.go

```diff
@@ -0,0 +1,225 @@
+package infix
+
+import (
+	"fmt"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+)
+
+type (
+	lexer struct {
+		input  string // expression being lex'ed
+		start  int    // start position of token
+		pos    int    // pos in the input
+		width  int    // width of last rune
+		tokens chan Tokval
+	}
+
+	stateFn func(*lexer) stateFn
+
+	// Tokval is the token type + value
+	Tokval struct {
+		T Token
+		V string
+	}
+)
+
+// eof mimics the EOF but for the input string.
+// The name isn't eoi to avoid confusion.
+const eof = -1
+
+func (t Tokval) String() string {
+	return fmt.Sprintf("Token(%s, %s)", t.T, t.V)
+}
+
+// Lex creates a concurrent lexer and returns a
+// channel of tokens processed from input.
+func Lex(input string) chan Tokval {
+	l := &lexer{
+		input:  input,
+		tokens: make(chan Tokval),
+	}
+
+	go l.run()
+	return l.tokens
+}
+
+// run the state machine
+func (l *lexer) run() {
```

Comment:
> I'm prone to not creating the goroutine because sometimes the user can use the main gorotine to run the loop,

But what's the advantages? This way is simple and the use is very straightforward:

```go
for tok := range Lex(src) {
    fmt.Printf("token: %s\n", tok)
}
```
The other way requires documentation to make clear that calling Run() and Tokens() on same goroutine deadlocks. I understand the benefits on more complex tasks but fail to see any point here.

> if all functions always creates goroutines internally the user has no choice and will have to lock the main goroutine on some kind of nil channel.

I dont get this. Only the producer ran on a new goroutine, the consumer can run anywhere you want.

Reply:
I was generalizing an idea, but I think this is not the best example to discuss it =P. Just forget about it